File: java-project-standard/assets/templates/adapters/src/main/java/__PKG_PATH__/adapters/ProviderFactory.java
package __PKG__.adapters;

import __PKG__.domain.Embedder;
import __PKG__.domain.Llm;

/**
 * Assembly seam: a config string selects an implementation. Default is mock; an unknown provider
 * fails loudly (never a silent fallback). This is the one place implementations are chosen.
 *
 * <p>Real backends (e.g. OpenAI) live in a gated sibling module (:adapters-openai) that is only
 * included when the {@code realProviders} Gradle property is set; unlock a branch below when it is.
 */
public final class ProviderFactory {
  private ProviderFactory() {}

  /**
   * Select an LLM implementation by provider id.
   *
   * @throws IllegalArgumentException when the provider is unknown (never silently falls back).
   */
  public static Llm makeLlm(String provider) {
    return switch (provider) {
      case "mock" -> new MockLlm();
      // case "openai" -> new OpenAiLlm(); // only with the gated :adapters-openai module
      default -> throw new IllegalArgumentException("unknown llmProvider: " + provider);
    };
  }

  /**
   * Select an Embedder implementation by provider id. Defaults to mock.
   *
   * @throws IllegalArgumentException when the provider is unknown.
   */
  public static Embedder makeEmbedder(String provider) {
    return switch (provider) {
      case "mock" -> new MockEmbedder();
      default -> throw new IllegalArgumentException("unknown embedder provider: " + provider);
    };
  }
}

File: android-project-standard/assets/templates/adapters/src/main/kotlin/__PKG_PATH__/adapters/ProviderFactory.kt
// 装配缝工厂:按 config 选具体实现。默认 mock;未知 provider 显式抛错,绝不沉默回退。
// :app 的 Hilt 组合根(AppModule)调用它注入 `Llm` / `Embedder`。
//
// 真实后端在 gated 的 `:adapters-openai` module 里;开启后在此 when 加一个分支(见 CLAUDE.md)。

package __APP_ID__.adapters

import __APP_ID__.domain.Embedder
import __APP_ID__.domain.Llm
import __APP_ID__.domain.ProviderError

/** provider 装配缝:config 字符串 → 实现。 */
public object ProviderFactory {
    /**
     * 按 provider 标识选 LLM 实现。
     *
     * @throws ProviderError.InvalidResponse 当 provider 未知(绝不沉默回退到默认)。
     */
    public fun makeLlm(provider: String): Llm =
        when (provider) {
            "mock" -> MockLlm()
            // "openai" -> OpenAiLlm()   // 仅当 gated :adapters-openai 启用时解开
            else -> throw ProviderError.InvalidResponse("unknown llmProvider: $provider")
        }

    /** 按 provider 标识选 Embedder 实现。默认 mock。 */
    public fun makeEmbedder(provider: String): Embedder =
        when (provider) {
            "mock" -> MockEmbedder()
            // "openai" -> OpenAiEmbedder()
            else -> throw ProviderError.InvalidResponse("unknown embedder provider: $provider")
        }
}
